subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/FileSystemWatchingBuildActionRunner.java

```diff
@@ -44,8 +48,12 @@ public FileSystemWatchingBuildActionRunner(BuildActionRunner delegate) {
     public Result run(BuildAction action, BuildController buildController) {
         GradleInternal gradle = buildController.getGradle();
         StartParameterInternal startParameter = gradle.getStartParameter();
-        BuildLifecycleAwareVirtualFileSystem virtualFileSystem = gradle.getServices().get(BuildLifecycleAwareVirtualFileSystem.class);
-        BuildOperationRunner buildOperationRunner = gradle.getServices().get(BuildOperationRunner.class);
+        ServiceRegistry services = gradle.getServices();
+        BuildLifecycleAwareVirtualFileSystem virtualFileSystem = services.get(BuildLifecycleAwareVirtualFileSystem.class);
+        StatStatistics.Collector statStatisticsCollector = services.get(StatStatistics.Collector.class);
+        FileHasherStatistics.Collector fileHasherStatisticsCollector = services.get(FileHasherStatistics.Collector.class);
+        DirectorySnapshotterStatistics.Collector directorySnapshotterStatisticsCollector = services.get(DirectorySnapshotterStatistics.Collector.class);
```

Comment:
How about the collectors implement a common interface, then you could obtain a `List<Collector>` here to obtain all registered collectors. We can also leave it like it is now and introduce the common interface when we introduce more collectors.

Reply:
I tried this, but for that to happen we'd need the `Collector` interface to be in a dependency subproject for all the projects. That leaves us with two options: a) reuse `:base-annotations` or b) add a new `:statistics` subproject.

I don't want to litter `:base-annotations` and adding a new subproject currently entails costs I wouldn't like to pay:

- our flat project hierarchy is already hard to handle with 116 subprojects
- adding one more JAR to our distribution is complicated
- adding one more JAR means we have a longer classpath, more JAR files for the classloader to parse etc.
- adding one more JAR means worsening the classpath length issues on Windows

I think once we introduced a hierarchical subproject directory structure, and started merging our JARs, then it would make sense to introduce even very small subprojects like  `:statistics`. Let's do this then.

Reply:
Right, we discussed this. 👍